tests/virt_autotest/virt_autotest_base.pm

```diff
@@ -85,29 +85,88 @@ sub generateXML {
+
+    $self->{assert_pattern}           = $assert_pattern;
+    $self->{add_junit_log_flag}       = $add_junit_log_flag;
+    $self->{upload_virt_log_flag}     = $upload_virt_log_flag;
+    $self->{log_dir}                  = $log_dir;
+    $self->{compressed_log_name}      = $compressed_log_name;
+    $self->{upload_guest_assets_flag} = $upload_guest_assets_flag;
+}
+
+#This is the subroutine called inside post_execute_script_run. It aims to do configurations have to be done after script
+#execution but before test assertion, for example, modifying boot options, communicating to peer with lock/unlcok and many other
+#things as long as they should be done and need to be done right after test execution or the following test steps may malfunction
+#without them. post_execute_script_configuration should be overriden in individual test modules.
+sub post_execute_script_configuration {
+    diag("You need to override this function post_execute_script_config in your test module");
+}
+
+#This is the subroutine called inside post_execute_script_run. This is introduced to faciliate tests that do not have or not convenient
+#to use assert pattern,  can not use directly returned output from execute_script_run or needs more customized way to manipuluate results.
+#So individual test modules can also have more control over how their test results should be extracted out, intead of just relying on a
+#singel assert pattern. post_execute_script_assertion needs to be overriden in individual test modules.
+sub post_execute_script_assertion {
+    diag("You need to override this function post_execute_script_assertion in your test module");
+}
+
+#This subroutine incorporates operations needs to be done after finishing executing script. It is further dividied into three parts,
+#including post_execute_script_configuration, upload_virt_logs and do test assertion. Test assertion can be done in two ways by using
+#assert pattern or newly introduced post_execute_script_assertion subroutine.
+sub post_execute_script_run {
+    my $self = shift;
+
+    $self->post_execute_script_configuration;
+
+    if ($self->{upload_virt_log_flag} eq "yes") {
+        upload_virt_logs($self->{log_dir}, $self->{compressed_log_name});
+    }
+    save_screenshot;
+
+    my $output = $self->{script_output};
+    if ($self->{assert_pattern}) {
+        unless ($output =~ /$self->{assert_pattern}/m) {
+            assert_script_run("grep -E \"$self->{assert_pattern}\" $self->{log_dir} -r || zcat /tmp/$self->{compressed_log_name}.tar.gz | grep -aE \"$self->{assert_pattern}\"");
+        }
+    }
+    else {
+        $self->post_execute_script_assertion;
+    }
+    save_screenshot;
+}
+
+#Adding junit log and uploading guest assets are wrapped up here in newly introduced subroutine post_run_test.
+#This is called after post_execute_script_run in run_test if test passes or in post_fail_hook if test fails.
+sub post_run_test {
+    my ($self, $script_output) = @_;
```

Comment:
Shall the second parameter be skipped, which is actually stored in $self->{script_output}? It can get better perl performance without passing large size string parameters.

Reply:
Let's keep this to make it looks more explicitly about what is going to be used and done.

Reply:
I am sorry. I do not think it is more readable this way. They are the same. When you apply the new way to store needed info in $self, then there should not be any need to pass the parameter in function calling, especially for our tests whose script output sometimes is very very long. I strongly suggest it.

Reply:
Done